// igniteui-angular/esm2022/lib/services/animation/angular-animation-player.mjs
import { EventEmitter } from '@angular/core';
export class IgxAngularAnimationPlayer {
    get position() {
        return this._innerPlayer.getPosition();
    }
    set position(value) {
        this.internalPlayer.setPosition(value);
    }
    constructor(internalPlayer) {
        this.internalPlayer = internalPlayer;
        this.animationStart = new EventEmitter();
        this.animationEnd = new EventEmitter();
        this.internalPlayer.onDone(() => this.onDone());
        const innerRenderer = this.internalPlayer._renderer;
        //  We need inner player as Angular.AnimationPlayer.getPosition returns always 0.
        // To workaround this we are getting the positions from the inner player.
        //  This is logged in Angular here - https://github.com/angular/angular/issues/18891
        //  As soon as this is resolved we can remove this hack
        this._innerPlayer = innerRenderer.engine.players[innerRenderer.engine.players.length - 1];
    }
    init() {
        this.internalPlayer.init();
    }
    play() {
        this.animationStart.emit({ owner: this });
        this.internalPlayer.play();
    }
    finish() {
        this.internalPlayer.finish();
        // TODO: when animation finish angular deletes all onDone handlers. Add handlers again if needed
    }
    reset() {
        this.internalPlayer.reset();
        // calling reset does not change hasStarted to false. This is why we are doing it here via internal field
        this.internalPlayer._started = false;
    }
    destroy() {
        this.internalPlayer.destroy();
    }
    hasStarted() {
        return this.internalPlayer.hasStarted();
    }
    onDone() {
        this.animationEnd.emit({ owner: this });
    }
}
