// igniteui-angular/esm2022/lib/directives/text-highlight/text-highlight.service.mjs
import { EventEmitter, Injectable } from '@angular/core';
import * as i0 from "@angular/core";
export class IgxTextHighlightService {
    constructor() {
        this.highlightGroupsMap = new Map();
        this.onActiveElementChanged = new EventEmitter();
    }
    /**
     * Activates the highlight at a given index.
     * (if such index exists)
     */
    setActiveHighlight(groupName, highlight) {
        this.highlightGroupsMap.set(groupName, highlight);
        this.onActiveElementChanged.emit(groupName);
    }
    /**
     * Clears any existing highlight.
     */
    clearActiveHighlight(groupName) {
        this.highlightGroupsMap.set(groupName, {
            index: -1
        });
        this.onActiveElementChanged.emit(groupName);
    }
    /**
     * Destroys a highlight group.
     */
    destroyGroup(groupName) {
        this.highlightGroupsMap.delete(groupName);
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.2.4", ngImport: i0, type: IgxTextHighlightService, deps: [], target: i0.ɵɵFactoryTarget.Injectable }); }
    static { this.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "17.2.4", ngImport: i0, type: IgxTextHighlightService, providedIn: 'root' }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.2.4", ngImport: i0, type: IgxTextHighlightService, decorators: [{
            type: Injectable,
            args: [{
                    providedIn: 'root'
                }]
        }], ctorParameters: () => [] });
