// igniteui-angular/esm2022/lib/grids/resizing/pivot-grid/pivot-resize-handle.directive.mjs
import { Directive, Input } from '@angular/core';
import { IgxResizeHandleDirective } from '../resize-handle.directive';
import * as i0 from "@angular/core";
import * as i1 from "./pivot-resizing.service";
/**
 * @hidden
 * @internal
 */
export class IgxPivotResizeHandleDirective extends IgxResizeHandleDirective {
    /**
     * @hidden
     */
    set pivotColumn(value) {
        this.column = value;
    }
    get pivotColumn() {
        return this.column;
    }
    constructor(zone, element, colResizingService) {
        super(zone, element, colResizingService);
        this.colResizingService = colResizingService;
    }
    /**
     * @hidden
     */
    onDoubleClick() {
        this._dblClick = true;
        this.initResizeService();
        this.rowHeaderGroup.grid.autoSizeRowDimension(this.rowHeaderGroup.parent.rootDimension);
    }
    /**
     * @hidden
     */
    initResizeService(event = null) {
        super.initResizeService(event);
        this.colResizingService.rowHeaderGroup = this.rowHeaderGroup;
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.2.4", ngImport: i0, type: IgxPivotResizeHandleDirective, deps: [{ token: i0.NgZone }, { token: i0.ElementRef }, { token: i1.IgxPivotColumnResizingService }], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "17.2.4", type: IgxPivotResizeHandleDirective, isStandalone: true, selector: "[igxPivotResizeHandle]", inputs: { pivotColumn: ["igxPivotResizeHandle", "pivotColumn"], rowHeaderGroup: ["igxPivotResizeHandleHeader", "rowHeaderGroup"] }, usesInheritance: true, ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.2.4", ngImport: i0, type: IgxPivotResizeHandleDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[igxPivotResizeHandle]',
                    standalone: true
                }]
        }], ctorParameters: () => [{ type: i0.NgZone }, { type: i0.ElementRef }, { type: i1.IgxPivotColumnResizingService }], propDecorators: { pivotColumn: [{
                type: Input,
                args: ['igxPivotResizeHandle']
            }], rowHeaderGroup: [{
                type: Input,
                args: ['igxPivotResizeHandleHeader']
            }] } });
