// igniteui-angular/schematics/utils/util.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.createHost = exports.ProjectType = void 0;
// from '@schematics/angular/utility/workspace-models'
var ProjectType;
(function (ProjectType) {
    ProjectType["Application"] = "application";
    ProjectType["Library"] = "library";
})(ProjectType || (exports.ProjectType = ProjectType = {}));
const createHost = (tree) => ({
    readFile: (path) => __awaiter(void 0, void 0, void 0, function* () {
        const data = tree.read(path);
        // can use fileBufferToString
        return data === null || data === void 0 ? void 0 : data.toString();
    }),
    writeFile: (path, data) => __awaiter(void 0, void 0, void 0, function* () {
        tree.overwrite(path, data);
    }),
    isDirectory: (path) => __awaiter(void 0, void 0, void 0, function* () { return !tree.exists(path) && tree.getDir(path).subfiles.length > 0; }),
    isFile: (path) => __awaiter(void 0, void 0, void 0, function* () { return tree.exists(path); })
});
exports.createHost = createHost;
