// igniteui-angular/esm2022/lib/tabs/tabs/tabs.directives.mjs
import { Directive } from '@angular/core';
import * as i0 from "@angular/core";
export class IgxTabHeaderLabelDirective {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.2.4", ngImport: i0, type: IgxTabHeaderLabelDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "17.2.4", type: IgxTabHeaderLabelDirective, isStandalone: true, selector: "igx-tab-header-label,[igxTabHeaderLabel]", ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.2.4", ngImport: i0, type: IgxTabHeaderLabelDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: 'igx-tab-header-label,[igxTabHeaderLabel]',
                    standalone: true
                }]
        }] });
export class IgxTabHeaderIconDirective {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.2.4", ngImport: i0, type: IgxTabHeaderIconDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "17.2.4", type: IgxTabHeaderIconDirective, isStandalone: true, selector: "igx-tab-header-icon,[igxTabHeaderIcon]", ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.2.4", ngImport: i0, type: IgxTabHeaderIconDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: 'igx-tab-header-icon,[igxTabHeaderIcon]',
                    standalone: true
                }]
        }] });
