// igniteui-angular/esm2022/lib/select/select-group.component.mjs
import { Component } from '@angular/core';
import { IgxDropDownGroupComponent } from '../drop-down/drop-down-group.component';
import * as i0 from "@angular/core";
/**
 * The `<igx-select-item>` is a container intended for row items in
 * a `<igx-select>` container.
 */
export class IgxSelectGroupComponent extends IgxDropDownGroupComponent {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.2.4", ngImport: i0, type: IgxSelectGroupComponent, deps: null, target: i0.ɵɵFactoryTarget.Component }); }
    static { this.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "17.2.4", type: IgxSelectGroupComponent, isStandalone: true, selector: "igx-select-item-group", usesInheritance: true, ngImport: i0, template: `
        <label id="{{labelId}}">{{ label }}</label>
        <ng-content select="igx-select-item"></ng-content>
    `, isInline: true }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.2.4", ngImport: i0, type: IgxSelectGroupComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'igx-select-item-group',
                    template: `
        <label id="{{labelId}}">{{ label }}</label>
        <ng-content select="igx-select-item"></ng-content>
    `,
                    standalone: true
                }]
        }] });
