// igniteui-angular/esm2022/lib/grids/summaries/grid-root-summary.pipe.mjs
import { Inject, Pipe } from '@angular/core';
import { IGX_GRID_BASE } from '../common/grid.interface';
import * as i0 from "@angular/core";
export class IgxSummaryDataPipe {
    constructor(grid) {
        this.grid = grid;
    }
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    transform(id, trigger = 0) {
        const summaryService = this.grid.summaryService;
        return summaryService.calculateSummaries(summaryService.rootSummaryID, this.grid.gridAPI.get_summary_data());
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.2.4", ngImport: i0, type: IgxSummaryDataPipe, deps: [{ token: IGX_GRID_BASE }], target: i0.ɵɵFactoryTarget.Pipe }); }
    static { this.ɵpipe = i0.ɵɵngDeclarePipe({ minVersion: "14.0.0", version: "17.2.4", ngImport: i0, type: IgxSummaryDataPipe, isStandalone: true, name: "igxGridSummaryDataPipe" }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.2.4", ngImport: i0, type: IgxSummaryDataPipe, decorators: [{
            type: Pipe,
            args: [{
                    name: 'igxGridSummaryDataPipe',
                    standalone: true
                }]
        }], ctorParameters: () => [{ type: undefined, decorators: [{
                    type: Inject,
                    args: [IGX_GRID_BASE]
                }] }] });
