// @igniteui/cli-core/update/package-resolve.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FEED_WEBCOMPONENTS_SPREADSHEET = exports.NPM_WEBCOMPONENTS_SPREADSHEET = exports.FEED_WEBCOMPONENTS_MAPS = exports.NPM_WEBCOMPONENTS_MAPS = exports.FEED_WEBCOMPONENTS_LAYOUTS = exports.NPM_WEBCOMPONENTS_LAYOUTS = exports.FEED_WEBCOMPONENTS_INPUTS = exports.NPM_WEBCOMPONENTS_INPUTS = exports.FEED_WEBCOMPONENTS_GRIDS = exports.NPM_WEBCOMPONENTS_GRIDS = exports.FEED_WEBCOMPONENTS_GAUGES = exports.NPM_WEBCOMPONENTS_GAUGES = exports.FEED_WEBCOMPONENTS_FDC3 = exports.NPM_WEBCOMPONENTS_FDC3 = exports.FEED_WEBCOMPONENTS_EXCEL = exports.NPM_WEBCOMPONENTS_EXCEL = exports.FEED_WEBCOMPONENTS_DATASOURCES = exports.NPM_WEBCOMPONENTS_DATASOURCES = exports.FEED_WEBCOMPONENTS_CORE = exports.NPM_WEBCOMPONENTS_CORE = exports.FEED_WEBCOMPONENTS_CHARTS = exports.NPM_WEBCOMPONENTS_CHARTS = exports.FEED_REACT_SPREADSHEET_CHART_ADAPTER = exports.NPM_REACT_SPREADSHEET_CHART_ADAPTER = exports.FEED_REACT_SPREADSHEET = exports.NPM_REACT_SPREADSHEET = exports.FEED_REACT_DATASOURCES = exports.NPM_REACT_DATASOURCES = exports.FEED_REACT_EXCEL = exports.NPM_REACT_EXCEL = exports.FEED_REACT_MAPS = exports.NPM_REACT_MAPS = exports.FEED_REACT_LAYOUTS = exports.NPM_REACT_LAYOUTS = exports.FEED_REACT_INPUTS = exports.NPM_REACT_INPUTS = exports.FEED_REACT_GRIDS = exports.NPM_REACT_GRIDS = exports.FEED_REACT_GAUGES = exports.NPM_REACT_GAUGES = exports.FEED_REACT_CHARTS = exports.NPM_REACT_CHARTS = exports.FEED_REACT_CORE = exports.NPM_REACT_CORE = exports.FEED_REACT = exports.NPM_REACT = exports.FEED_DOCK_MANAGER = exports.NPM_DOCK_MANAGER = exports.FEED_ANGULAR = exports.NPM_ANGULAR = void 0;
exports.getUpgradeablePackages = exports.resolvePackage = exports.UPGRADEABLE_PACKAGES = exports.FEED_WEBCOMPONENTS_CHART_ADAPTER = exports.NPM_WEBCOMPONENTS_CHART_ADAPTER = void 0;
const types_1 = require("../types");
const util_1 = require("../util");
// angular
exports.NPM_ANGULAR = "igniteui-angular";
exports.FEED_ANGULAR = "@infragistics/igniteui-angular";
// dock manager
exports.NPM_DOCK_MANAGER = "igniteui-dockmanager";
exports.FEED_DOCK_MANAGER = "@infragistics/igniteui-dockmanager";
// react
exports.NPM_REACT = "igniteui-react";
exports.FEED_REACT = "@infragistics/igniteui-react";
exports.NPM_REACT_CORE = "igniteui-react-core";
exports.FEED_REACT_CORE = "@infragistics/igniteui-react-core";
exports.NPM_REACT_CHARTS = "igniteui-react-charts";
exports.FEED_REACT_CHARTS = "@infragistics/igniteui-react-charts";
exports.NPM_REACT_GAUGES = "igniteui-react-gauges";
exports.FEED_REACT_GAUGES = "@infragistics/igniteui-react-gauges";
exports.NPM_REACT_GRIDS = "igniteui-react-grids";
exports.FEED_REACT_GRIDS = "@infragistics/igniteui-react-grids";
exports.NPM_REACT_INPUTS = "igniteui-react-inputs";
exports.FEED_REACT_INPUTS = "@infragistics/igniteui-react-inputs";
exports.NPM_REACT_LAYOUTS = "igniteui-react-layouts";
exports.FEED_REACT_LAYOUTS = "@infragistics/igniteui-react-layouts";
exports.NPM_REACT_MAPS = "igniteui-react-maps";
exports.FEED_REACT_MAPS = "@infragistics/igniteui-react-maps";
exports.NPM_REACT_EXCEL = "igniteui-react-excel";
exports.FEED_REACT_EXCEL = "@infragistics/igniteui-react-excel";
exports.NPM_REACT_DATASOURCES = "igniteui-react-datasources";
exports.FEED_REACT_DATASOURCES = "@infragistics/igniteui-react-datasources";
exports.NPM_REACT_SPREADSHEET = "igniteui-react-spreadsheet";
exports.FEED_REACT_SPREADSHEET = "@infragistics/igniteui-react-spreadsheet";
exports.NPM_REACT_SPREADSHEET_CHART_ADAPTER = "igniteui-react-spreadsheet-chart-adapter";
exports.FEED_REACT_SPREADSHEET_CHART_ADAPTER = "@infragistics/igniteui-react-spreadsheet-chart-adapter";
// webcomponents
exports.NPM_WEBCOMPONENTS_CHARTS = "igniteui-webcomponents-charts";
exports.FEED_WEBCOMPONENTS_CHARTS = "@infragistics/igniteui-webcomponents-charts";
exports.NPM_WEBCOMPONENTS_CORE = "igniteui-webcomponents-core";
exports.FEED_WEBCOMPONENTS_CORE = "@infragistics/igniteui-webcomponents-core";
exports.NPM_WEBCOMPONENTS_DATASOURCES = "igniteui-webcomponents-datasources";
exports.FEED_WEBCOMPONENTS_DATASOURCES = "@infragistics/igniteui-webcomponents-datasources";
exports.NPM_WEBCOMPONENTS_EXCEL = "igniteui-webcomponents-excel";
exports.FEED_WEBCOMPONENTS_EXCEL = "@infragistics/igniteui-webcomponents-excel";
exports.NPM_WEBCOMPONENTS_FDC3 = "igniteui-webcomponents-fdc3";
exports.FEED_WEBCOMPONENTS_FDC3 = "@infragistics/igniteui-webcomponents-fdc3";
exports.NPM_WEBCOMPONENTS_GAUGES = "igniteui-webcomponents-gauges";
exports.FEED_WEBCOMPONENTS_GAUGES = "@infragistics/igniteui-webcomponents-gauges";
exports.NPM_WEBCOMPONENTS_GRIDS = "igniteui-webcomponents-grids";
exports.FEED_WEBCOMPONENTS_GRIDS = "@infragistics/igniteui-webcomponents-grids";
exports.NPM_WEBCOMPONENTS_INPUTS = "igniteui-webcomponents-inputs";
exports.FEED_WEBCOMPONENTS_INPUTS = "@infragistics/igniteui-webcomponents-inputs";
exports.NPM_WEBCOMPONENTS_LAYOUTS = "igniteui-webcomponents-layouts";
exports.FEED_WEBCOMPONENTS_LAYOUTS = "@infragistics/igniteui-webcomponents-layouts";
exports.NPM_WEBCOMPONENTS_MAPS = "igniteui-webcomponents-maps";
exports.FEED_WEBCOMPONENTS_MAPS = "@infragistics/igniteui-webcomponents-maps";
exports.NPM_WEBCOMPONENTS_SPREADSHEET = "igniteui-webcomponents-spreadsheet";
exports.FEED_WEBCOMPONENTS_SPREADSHEET = "@infragistics/igniteui-webcomponents-spreadsheet";
exports.NPM_WEBCOMPONENTS_CHART_ADAPTER = "igniteui-webcomponents-spreadsheet-chart-adapter";
exports.FEED_WEBCOMPONENTS_CHART_ADAPTER = "@infragistics/igniteui-webcomponents-spreadsheet-chart-adapter";
/** packages map of `trial: licensed` names */
exports.UPGRADEABLE_PACKAGES = {
    // angular
    [exports.NPM_ANGULAR]: exports.FEED_ANGULAR,
    // dock manager
    [exports.NPM_DOCK_MANAGER]: exports.FEED_DOCK_MANAGER,
    // react
    [exports.NPM_REACT]: exports.FEED_REACT,
    [exports.NPM_REACT_CORE]: exports.FEED_REACT_CORE,
    [exports.NPM_REACT_CHARTS]: exports.FEED_REACT_CHARTS,
    [exports.NPM_REACT_GAUGES]: exports.FEED_REACT_GAUGES,
    [exports.NPM_REACT_GRIDS]: exports.FEED_REACT_GRIDS,
    [exports.NPM_REACT_INPUTS]: exports.FEED_REACT_INPUTS,
    [exports.NPM_REACT_LAYOUTS]: exports.FEED_REACT_LAYOUTS,
    [exports.NPM_REACT_MAPS]: exports.FEED_REACT_MAPS,
    [exports.NPM_REACT_EXCEL]: exports.FEED_REACT_EXCEL,
    [exports.NPM_REACT_DATASOURCES]: exports.FEED_REACT_DATASOURCES,
    [exports.NPM_REACT_SPREADSHEET]: exports.FEED_REACT_SPREADSHEET,
    [exports.NPM_REACT_SPREADSHEET_CHART_ADAPTER]: exports.FEED_REACT_SPREADSHEET_CHART_ADAPTER,
    // webcomponents
    [exports.NPM_WEBCOMPONENTS_CHARTS]: exports.FEED_WEBCOMPONENTS_CHARTS,
    [exports.NPM_WEBCOMPONENTS_CORE]: exports.FEED_WEBCOMPONENTS_CORE,
    [exports.NPM_WEBCOMPONENTS_DATASOURCES]: exports.FEED_WEBCOMPONENTS_DATASOURCES,
    [exports.NPM_WEBCOMPONENTS_EXCEL]: exports.FEED_WEBCOMPONENTS_EXCEL,
    [exports.NPM_WEBCOMPONENTS_FDC3]: exports.FEED_WEBCOMPONENTS_FDC3,
    [exports.NPM_WEBCOMPONENTS_GAUGES]: exports.FEED_WEBCOMPONENTS_GAUGES,
    [exports.NPM_WEBCOMPONENTS_GRIDS]: exports.FEED_WEBCOMPONENTS_GRIDS,
    [exports.NPM_WEBCOMPONENTS_INPUTS]: exports.FEED_WEBCOMPONENTS_INPUTS,
    [exports.NPM_WEBCOMPONENTS_LAYOUTS]: exports.FEED_WEBCOMPONENTS_LAYOUTS,
    [exports.NPM_WEBCOMPONENTS_MAPS]: exports.FEED_WEBCOMPONENTS_MAPS,
    [exports.NPM_WEBCOMPONENTS_SPREADSHEET]: exports.FEED_WEBCOMPONENTS_SPREADSHEET,
    [exports.NPM_WEBCOMPONENTS_CHART_ADAPTER]: exports.FEED_WEBCOMPONENTS_CHART_ADAPTER
};
function resolvePackage(packageName) {
    const fs = util_1.App.container.get(types_1.FS_TOKEN);
    // read project package JSON
    if (fs.fileExists("./package.json")) {
        const fileContent = fs.readFile("./package.json");
        const packageJson = JSON.parse(fileContent);
        const dependencies = packageJson["dependencies"];
        const licensed = exports.UPGRADEABLE_PACKAGES[packageName];
        if (dependencies[licensed]) {
            return licensed;
        }
    }
    return packageName;
}
exports.resolvePackage = resolvePackage;
function getUpgradeablePackages() {
    const fs = util_1.App.container.get(types_1.FS_TOKEN);
    const upgradeable = [];
    if (fs.fileExists("./package.json")) {
        const fileContent = fs.readFile("./package.json");
        const packageJson = JSON.parse(fileContent);
        const dependencies = packageJson["dependencies"];
        for (const packageEntry in exports.UPGRADEABLE_PACKAGES) {
            if (dependencies[packageEntry]) {
                upgradeable.push({
                    trial: packageEntry,
                    licensed: exports.UPGRADEABLE_PACKAGES[packageEntry]
                });
            }
        }
    }
    return upgradeable;
}
exports.getUpgradeablePackages = getUpgradeablePackages;
