// igniteui-angular/esm2022/lib/progressbar/public_api.mjs
import { IgxProgressBarGradientDirective, IgxProgressBarTextTemplateDirective } from './progressbar.common';
import { IgxCircularProgressBarComponent, IgxLinearProgressBarComponent } from './progressbar.component';
export * from './progressbar.common';
export { IgxTextAlign, IgxProgressType, IgxLinearProgressBarComponent, IgxCircularProgressBarComponent } from './progressbar.component';
/* NOTE: Progress bar (linear and circular) directives collection for ease-of-use import in standalone components scenario */
export const IGX_PROGRESS_BAR_DIRECTIVES = [
    IgxLinearProgressBarComponent,
    IgxCircularProgressBarComponent,
    IgxProgressBarTextTemplateDirective,
    IgxProgressBarGradientDirective
];
/* NOTE: Linear progress bar directives collection for ease-of-use import in standalone components scenario */
export const IGX_LINEAR_PROGRESS_BAR_DIRECTIVES = [
    IgxLinearProgressBarComponent,
    IgxProgressBarGradientDirective
];
/* NOTE: Circular progress bar directives collection for ease-of-use import in standalone components scenario */
export const IGX_CIRCULAR_PROGRESS_BAR_DIRECTIVES = [
    IgxCircularProgressBarComponent,
    IgxProgressBarTextTemplateDirective,
    IgxProgressBarGradientDirective
];
