// igniteui-angular/esm2022/lib/stepper/stepper.service.mjs
import { Injectable } from '@angular/core';
import { IgxStepperOrientation } from './stepper.common';
import * as i0 from "@angular/core";
/** @hidden @internal */
export class IgxStepperService {
    constructor() {
        this.collapsingSteps = new Set();
        this.linearDisabledSteps = new Set();
        this.visitedSteps = new Set();
    }
    /**
     * Activates the step, fires the steps change event and plays animations.
     */
    expand(step) {
        if (this.activeStep === step) {
            return;
        }
        const cancel = this.emitActivatingEvent(step);
        if (cancel) {
            return;
        }
        this.collapsingSteps.delete(step);
        this.previousActiveStep = this.activeStep;
        this.activeStep = step;
        this.activeStep.activeChange.emit(true);
        this.collapsingSteps.add(this.previousActiveStep);
        this.visitedSteps.add(this.activeStep);
        if (this.stepper.orientation === IgxStepperOrientation.Vertical) {
            this.previousActiveStep.playCloseAnimation(this.previousActiveStep.contentContainer);
            this.activeStep.cdr.detectChanges();
            this.activeStep.playOpenAnimation(this.activeStep.contentContainer);
        }
        else {
            this.activeStep.cdr.detectChanges();
            this.stepper.playHorizontalAnimations();
        }
    }
    /**
     * Activates the step and fires the steps change event without playing animations.
     */
    expandThroughApi(step) {
        if (this.activeStep === step) {
            return;
        }
        this.collapsingSteps.clear();
        this.previousActiveStep = this.activeStep;
        this.activeStep = step;
        if (this.previousActiveStep) {
            this.previousActiveStep.tabIndex = -1;
        }
        this.activeStep.tabIndex = 0;
        this.visitedSteps.add(this.activeStep);
        this.activeStep.cdr.markForCheck();
        this.previousActiveStep?.cdr.markForCheck();
        this.activeStep.activeChange.emit(true);
        this.previousActiveStep?.activeChange.emit(false);
    }
    /**
     * Collapses the currently active step and fires the change event.
     */
    collapse(step) {
        if (this.activeStep === step) {
            return;
        }
        step.activeChange.emit(false);
        this.collapsingSteps.delete(step);
    }
    /**
     * Determines the steps that should be marked as visited based on the active step.
     */
    calculateVisitedSteps() {
        this.stepper.steps.forEach(step => {
            if (step.index <= this.activeStep.index) {
                this.visitedSteps.add(step);
            }
            else {
                this.visitedSteps.delete(step);
            }
        });
    }
    /**
     * Determines the steps that should be disabled in linear mode based on the validity of the active step.
     */
    calculateLinearDisabledSteps() {
        if (!this.activeStep) {
            return;
        }
        if (this.activeStep.isValid) {
            const firstRequiredIndex = this.getNextRequiredStep();
            if (firstRequiredIndex !== -1) {
                this.updateLinearDisabledSteps(firstRequiredIndex);
            }
            else {
                this.linearDisabledSteps.clear();
            }
        }
        else {
            this.stepper.steps.forEach(s => {
                if (s.index > this.activeStep.index) {
                    this.linearDisabledSteps.add(s);
                }
            });
        }
    }
    emitActivatingEvent(step) {
        const args = {
            owner: this.stepper,
            newIndex: step.index,
            oldIndex: this.activeStep.index,
            cancel: false
        };
        this.stepper.activeStepChanging.emit(args);
        return args.cancel;
    }
    /**
     * Updates the linearDisabled steps from the current active step to the next required invalid step.
     *
     * @param toIndex the index of the last step that should be enabled.
     */
    updateLinearDisabledSteps(toIndex) {
        this.stepper.steps.forEach(s => {
            if (s.index > this.activeStep.index) {
                if (s.index <= toIndex) {
                    this.linearDisabledSteps.delete(s);
                }
                else {
                    this.linearDisabledSteps.add(s);
                }
            }
        });
    }
    getNextRequiredStep() {
        if (!this.activeStep) {
            return;
        }
        return this.stepper.steps.findIndex(s => s.index > this.activeStep.index && !s.optional && !s.disabled && !s.isValid);
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.2.4", ngImport: i0, type: IgxStepperService, deps: [], target: i0.ɵɵFactoryTarget.Injectable }); }
    static { this.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "17.2.4", ngImport: i0, type: IgxStepperService }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.2.4", ngImport: i0, type: IgxStepperService, decorators: [{
            type: Injectable
        }] });
