// igniteui-angular/esm2022/lib/services/theme/theme.service.mjs
import { Inject, Injectable, Input } from '@angular/core';
import { mkenum } from '../../core/utils';
import { Subject } from 'rxjs';
import { DOCUMENT } from '@angular/common';
import * as i0 from "@angular/core";
const Theme = /*@__PURE__*/ mkenum({
    Material: 'material',
    Fluent: 'fluent',
    Bootstrap: 'bootstrap',
    IndigoDesign: 'indigo-design'
});
export class ThemeService {
    /**
     * Sets the theme of the component.
     * Allowed values of type IgxTheme.
     */
    set theme(value) {
        this._theme = value;
    }
    /**
     * Returns the theme of the component.
     * The returned value is of type IgxTheme.
     */
    get theme() {
        return this._theme;
    }
    constructor(document) {
        this.document = document;
        this._theme$ = new Subject();
        this._subscription = this._theme$.asObservable().subscribe(value => {
            this._theme = value;
        });
    }
    getCssProp(element) {
        if (!this._theme) {
            const cssProp = this.document.defaultView
                .getComputedStyle(element.nativeElement)
                .getPropertyValue('--theme')
                .trim();
            if (cssProp !== '') {
                Promise.resolve().then(() => {
                    this._theme$.next(cssProp);
                });
            }
        }
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.2.4", ngImport: i0, type: ThemeService, deps: [{ token: DOCUMENT }], target: i0.ɵɵFactoryTarget.Injectable }); }
    static { this.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "17.2.4", ngImport: i0, type: ThemeService, providedIn: 'root' }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.2.4", ngImport: i0, type: ThemeService, decorators: [{
            type: Injectable,
            args: [{
                    providedIn: 'root'
                }]
        }], ctorParameters: () => [{ type: undefined, decorators: [{
                    type: Inject,
                    args: [DOCUMENT]
                }] }], propDecorators: { theme: [{
                type: Input
            }] } });
