// igniteui-angular/esm2022/lib/grids/filtering/excel-style/common.mjs
import { FilteringExpressionsTree } from '../../../data-operations/filtering-expressions-tree';
/**
 * @hidden @internal
 */
export class FilterListItem {
    constructor() {
        this.isSpecial = false;
        this.isBlanks = false;
    }
}
/**
 * @hidden
 */
export class ExpressionUI {
    constructor() {
        this.isSelected = false;
        this.isVisible = true;
    }
}
/**
 * @hidden @internal
 */
export class ActiveElement {
}
export function generateExpressionsList(expressions, operator, expressionsUIs) {
    generateExpressionsListRecursive(expressions, operator, expressionsUIs);
    // The beforeOperator of the first expression and the afterOperator of the last expression should be null
    if (expressionsUIs.length) {
        expressionsUIs[expressionsUIs.length - 1].afterOperator = null;
    }
}
function generateExpressionsListRecursive(expressions, operator, expressionsUIs) {
    if (!expressions) {
        return;
    }
    if (expressions instanceof FilteringExpressionsTree) {
        const expressionsTree = expressions;
        for (const operand of expressionsTree.filteringOperands) {
            generateExpressionsListRecursive(operand, expressionsTree.operator, expressionsUIs);
        }
        if (expressionsUIs.length) {
            expressionsUIs[expressionsUIs.length - 1].afterOperator = operator;
        }
    }
    else {
        const exprUI = new ExpressionUI();
        exprUI.expression = expressions;
        exprUI.afterOperator = operator;
        const prevExprUI = expressionsUIs[expressionsUIs.length - 1];
        if (prevExprUI) {
            exprUI.beforeOperator = prevExprUI.afterOperator;
        }
        expressionsUIs.push(exprUI);
    }
}
