// igniteui-angular/esm2022/animations/util.mjs
import { fadeIn, fadeOut } from './fade';
import { flipBottom, flipHorBck, flipHorFwd, flipLeft, flipRight, flipTop, flipVerBck, flipVerFwd } from './flip';
import { growVerIn, growVerOut } from './grow';
import { blink, heartbeat, pulsateBck, pulsateFwd, shakeBl, shakeBottom, shakeBr, shakeCenter, shakeHor, shakeLeft, shakeRight, shakeTl, shakeTop, shakeTr, shakeVer } from './misc';
import { rotateInBl, rotateInBottom, rotateInBr, rotateInCenter, rotateInDiagonal1, rotateInDiagonal2, rotateInHor, rotateInLeft, rotateInRight, rotateInTl, rotateInTop, rotateInTr, rotateInVer, rotateOutBl, rotateOutBottom, rotateOutBr, rotateOutCenter, rotateOutDiagonal1, rotateOutDiagonal2, rotateOutHor, rotateOutLeft, rotateOutRight, rotateOutTl, rotateOutTop, rotateOutTr, rotateOutVer } from './rotate';
import { scaleInBl, scaleInBottom, scaleInBr, scaleInCenter, scaleInHorCenter, scaleInHorLeft, scaleInHorRight, scaleInLeft, scaleInRight, scaleInTl, scaleInTop, scaleInTr, scaleInVerBottom, scaleInVerCenter, scaleInVerTop, scaleOutBl, scaleOutBottom, scaleOutBr, scaleOutCenter, scaleOutHorCenter, scaleOutHorLeft, scaleOutHorRight, scaleOutLeft, scaleOutRight, scaleOutTl, scaleOutTop, scaleOutTr, scaleOutVerBottom, scaleOutVerCenter, scaleOutVerTop } from './scale';
import { slideInTop, slideInBottom, slideOutTop, slideOutBottom, slideInRight, slideInLeft, slideOutRight, slideOutLeft, slideInTr, slideInBl, slideOutTr, slideOutBl, slideInBr, slideInTl, slideOutBr, slideOutTl } from './slide';
import { swingInTopFwd, swingInBottomFwd, swingOutTopFwd, swingOutBottomFwd, swingInRightFwd, swingInLeftFwd, swingOutRightFwd, swingOutLefttFwd, swingInTopBck, swingInBottomBck, swingOutTopBck, swingOutBottomBck, swingInRightBck, swingInLeftBck, swingOutRightBck, swingOutLeftBck } from './swing';
export class AnimationUtil {
    constructor() {
        this.oppositeAnimation = new Map([
            [fadeIn, fadeIn],
            [fadeOut, fadeOut],
            [flipTop, flipBottom],
            [flipBottom, flipTop],
            [flipRight, flipLeft],
            [flipLeft, flipRight],
            [flipHorFwd, flipHorBck],
            [flipHorBck, flipHorFwd],
            [flipVerFwd, flipVerBck],
            [flipVerBck, flipVerFwd],
            [growVerIn, growVerIn],
            [growVerOut, growVerOut],
            [heartbeat, heartbeat],
            [pulsateFwd, pulsateBck],
            [pulsateBck, pulsateFwd],
            [blink, blink],
            [shakeHor, shakeHor],
            [shakeVer, shakeVer],
            [shakeTop, shakeTop],
            [shakeBottom, shakeBottom],
            [shakeRight, shakeRight],
            [shakeLeft, shakeLeft],
            [shakeCenter, shakeCenter],
            [shakeTr, shakeTr],
            [shakeBr, shakeBr],
            [shakeBl, shakeBl],
            [shakeTl, shakeTl],
            [rotateInCenter, rotateInCenter],
            [rotateOutCenter, rotateOutCenter],
            [rotateInTop, rotateInBottom],
            [rotateOutTop, rotateOutBottom],
            [rotateInRight, rotateInLeft],
            [rotateOutRight, rotateOutLeft],
            [rotateInLeft, rotateInRight],
            [rotateOutLeft, rotateOutRight],
            [rotateInBottom, rotateInTop],
            [rotateOutBottom, rotateOutTop],
            [rotateInTr, rotateInBl],
            [rotateOutTr, rotateOutBl],
            [rotateInBr, rotateInTl],
            [rotateOutBr, rotateOutTl],
            [rotateInBl, rotateInTr],
            [rotateOutBl, rotateOutTr],
            [rotateInTl, rotateInBr],
            [rotateOutTl, rotateOutBr],
            [rotateInDiagonal1, rotateInDiagonal1],
            [rotateOutDiagonal1, rotateOutDiagonal1],
            [rotateInDiagonal2, rotateInDiagonal2],
            [rotateOutDiagonal2, rotateOutDiagonal2],
            [rotateInHor, rotateInHor],
            [rotateOutHor, rotateOutHor],
            [rotateInVer, rotateInVer],
            [rotateOutVer, rotateOutVer],
            [scaleInTop, scaleInBottom],
            [scaleOutTop, scaleOutBottom],
            [scaleInRight, scaleInLeft],
            [scaleOutRight, scaleOutLeft],
            [scaleInBottom, scaleInTop],
            [scaleOutBottom, scaleOutTop],
            [scaleInLeft, scaleInRight],
            [scaleOutLeft, scaleOutRight],
            [scaleInCenter, scaleInCenter],
            [scaleOutCenter, scaleOutCenter],
            [scaleInTr, scaleInBl],
            [scaleOutTr, scaleOutBl],
            [scaleInBr, scaleInTl],
            [scaleOutBr, scaleOutTl],
            [scaleInBl, scaleInTr],
            [scaleOutBl, scaleOutTr],
            [scaleInTl, scaleInBr],
            [scaleOutTl, scaleOutBr],
            [scaleInVerTop, scaleInVerBottom],
            [scaleOutVerTop, scaleOutVerBottom],
            [scaleInVerBottom, scaleInVerTop],
            [scaleOutVerBottom, scaleOutVerTop],
            [scaleInVerCenter, scaleInVerCenter],
            [scaleOutVerCenter, scaleOutVerCenter],
            [scaleInHorCenter, scaleInHorCenter],
            [scaleOutHorCenter, scaleOutHorCenter],
            [scaleInHorLeft, scaleInHorRight],
            [scaleOutHorLeft, scaleOutHorRight],
            [scaleInHorRight, scaleInHorLeft],
            [scaleOutHorRight, scaleOutHorLeft],
            [slideInTop, slideInBottom],
            [slideOutTop, slideOutBottom],
            [slideInRight, slideInLeft],
            [slideOutRight, slideOutLeft],
            [slideInBottom, slideInTop],
            [slideOutBottom, slideOutTop],
            [slideInLeft, slideInRight],
            [slideOutLeft, slideOutRight],
            [slideInTr, slideInBl],
            [slideOutTr, slideOutBl],
            [slideInBr, slideInTl],
            [slideOutBr, slideOutTl],
            [slideInBl, slideInTr],
            [slideOutBl, slideOutTr],
            [slideInTl, slideInBr],
            [slideOutTl, slideOutBr],
            [swingInTopFwd, swingInBottomFwd],
            [swingOutTopFwd, swingOutBottomFwd],
            [swingInRightFwd, swingInLeftFwd],
            [swingOutRightFwd, swingOutLefttFwd],
            [swingInLeftFwd, swingInRightFwd],
            [swingOutLefttFwd, swingOutRightFwd],
            [swingInBottomFwd, swingInTopFwd],
            [swingOutBottomFwd, swingOutTopFwd],
            [swingInTopBck, swingInBottomBck],
            [swingOutTopBck, swingOutBottomBck],
            [swingInRightBck, swingInLeftBck],
            [swingOutRightBck, swingOutLeftBck],
            [swingInBottomBck, swingInTopBck],
            [swingOutBottomBck, swingOutTopBck],
            [swingInLeftBck, swingInRightBck],
            [swingOutLeftBck, swingOutRightBck],
        ]);
        this.horizontalAnimations = [
            flipRight,
            flipLeft,
            flipVerFwd,
            flipVerBck,
            rotateInRight,
            rotateOutRight,
            rotateInLeft,
            rotateOutLeft,
            rotateInTr,
            rotateOutTr,
            rotateInBr,
            rotateOutBr,
            rotateInBl,
            rotateOutBl,
            rotateInTl,
            rotateOutTl,
            scaleInRight,
            scaleOutRight,
            scaleInLeft,
            scaleOutLeft,
            scaleInTr,
            scaleOutTr,
            scaleInBr,
            scaleOutBr,
            scaleInBl,
            scaleOutBl,
            scaleInTl,
            scaleOutTl,
            scaleInHorLeft,
            scaleOutHorLeft,
            scaleInHorRight,
            scaleOutHorRight,
            slideInRight,
            slideOutRight,
            slideInLeft,
            slideOutLeft,
            slideInTr,
            slideOutTr,
            slideInBr,
            slideOutBr,
            slideInBl,
            slideOutBl,
            slideInTl,
            slideOutTl,
            swingInRightFwd,
            swingOutRightFwd,
            swingInLeftFwd,
            swingOutLefttFwd,
            swingInRightBck,
            swingOutRightBck,
            swingInLeftBck,
            swingOutLeftBck,
        ];
        this.verticalAnimations = [
            flipTop,
            flipBottom,
            flipHorFwd,
            flipHorBck,
            growVerIn,
            growVerOut,
            rotateInTop,
            rotateOutTop,
            rotateInBottom,
            rotateOutBottom,
            rotateInTr,
            rotateOutTr,
            rotateInBr,
            rotateOutBr,
            rotateInBl,
            rotateOutBl,
            rotateInTl,
            rotateOutTl,
            scaleInTop,
            scaleOutTop,
            scaleInBottom,
            scaleOutBottom,
            scaleInTr,
            scaleOutTr,
            scaleInBr,
            scaleOutBr,
            scaleInBl,
            scaleOutBl,
            scaleInTl,
            scaleOutTl,
            scaleInVerTop,
            scaleOutVerTop,
            scaleInVerBottom,
            scaleOutVerBottom,
            slideInTop,
            slideOutTop,
            slideInBottom,
            slideOutBottom,
            slideInTr,
            slideOutTr,
            slideInBr,
            slideOutBr,
            slideInBl,
            slideOutBl,
            slideInTl,
            slideOutTl,
            swingInTopFwd,
            swingOutTopFwd,
            swingInBottomFwd,
            swingOutBottomFwd,
            swingInTopBck,
            swingOutTopBck,
            swingInBottomBck,
            swingOutBottomBck,
        ];
    }
    static instance() {
        return this._instance || (this._instance = new this());
    }
    reverseAnimationResolver(animation) {
        return this.oppositeAnimation.get(animation) ?? animation;
    }
    isHorizontalAnimation(animation) {
        return this.horizontalAnimations.includes(animation);
    }
    isVerticalAnimation(animation) {
        return this.verticalAnimations.includes(animation);
    }
}
