// igniteui-angular/esm2022/lib/grids/pivot-grid/pivot-grid.directives.mjs
import { Directive } from '@angular/core';
import * as i0 from "@angular/core";
/**
 * @hidden
 */
export class IgxPivotValueChipTemplateDirective {
    constructor(template) {
        this.template = template;
    }
    static ngTemplateContextGuard(_directive, context) {
        return true;
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.2.4", ngImport: i0, type: IgxPivotValueChipTemplateDirective, deps: [{ token: i0.TemplateRef }], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "17.2.4", type: IgxPivotValueChipTemplateDirective, isStandalone: true, selector: "[igxPivotValueChip]", ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.2.4", ngImport: i0, type: IgxPivotValueChipTemplateDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[igxPivotValueChip]',
                    standalone: true
                }]
        }], ctorParameters: () => [{ type: i0.TemplateRef }] });
