// igniteui-angular/esm2022/lib/grids/columns/validators.directive.mjs
import { Directive } from '@angular/core';
import { RequiredValidator, NG_VALIDATORS, MinValidator, MaxValidator, EmailValidator, MinLengthValidator, MaxLengthValidator, PatternValidator } from '@angular/forms';
import * as i0 from "@angular/core";
export class IgxColumnRequiredValidatorDirective extends RequiredValidator {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.2.4", ngImport: i0, type: IgxColumnRequiredValidatorDirective, deps: null, target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "17.2.4", type: IgxColumnRequiredValidatorDirective, isStandalone: true, selector: "igx-column[required]", providers: [{
                provide: NG_VALIDATORS,
                useExisting: IgxColumnRequiredValidatorDirective,
                multi: true
            }], usesInheritance: true, ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.2.4", ngImport: i0, type: IgxColumnRequiredValidatorDirective, decorators: [{
            type: Directive,
            args: [{
                    // eslint-disable-next-line @angular-eslint/directive-selector
                    selector: 'igx-column[required]',
                    providers: [{
                            provide: NG_VALIDATORS,
                            useExisting: IgxColumnRequiredValidatorDirective,
                            multi: true
                        }],
                    standalone: true
                }]
        }] });
export class IgxColumnMinValidatorDirective extends MinValidator {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.2.4", ngImport: i0, type: IgxColumnMinValidatorDirective, deps: null, target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "17.2.4", type: IgxColumnMinValidatorDirective, isStandalone: true, selector: "igx-column[min]", providers: [{
                provide: NG_VALIDATORS,
                useExisting: IgxColumnMinValidatorDirective,
                multi: true
            }], usesInheritance: true, ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.2.4", ngImport: i0, type: IgxColumnMinValidatorDirective, decorators: [{
            type: Directive,
            args: [{
                    // eslint-disable-next-line @angular-eslint/directive-selector
                    selector: 'igx-column[min]',
                    providers: [{
                            provide: NG_VALIDATORS,
                            useExisting: IgxColumnMinValidatorDirective,
                            multi: true
                        }],
                    standalone: true
                }]
        }] });
export class IgxColumnMaxValidatorDirective extends MaxValidator {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.2.4", ngImport: i0, type: IgxColumnMaxValidatorDirective, deps: null, target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "17.2.4", type: IgxColumnMaxValidatorDirective, isStandalone: true, selector: "igx-column[max]", providers: [{
                provide: NG_VALIDATORS,
                useExisting: IgxColumnMaxValidatorDirective,
                multi: true
            }], usesInheritance: true, ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.2.4", ngImport: i0, type: IgxColumnMaxValidatorDirective, decorators: [{
            type: Directive,
            args: [{
                    // eslint-disable-next-line @angular-eslint/directive-selector
                    selector: 'igx-column[max]',
                    providers: [{
                            provide: NG_VALIDATORS,
                            useExisting: IgxColumnMaxValidatorDirective,
                            multi: true
                        }],
                    standalone: true
                }]
        }] });
export class IgxColumnEmailValidatorDirective extends EmailValidator {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.2.4", ngImport: i0, type: IgxColumnEmailValidatorDirective, deps: null, target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "17.2.4", type: IgxColumnEmailValidatorDirective, isStandalone: true, selector: "igx-column[email]", providers: [{
                provide: NG_VALIDATORS,
                useExisting: IgxColumnEmailValidatorDirective,
                multi: true
            }], usesInheritance: true, ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.2.4", ngImport: i0, type: IgxColumnEmailValidatorDirective, decorators: [{
            type: Directive,
            args: [{
                    // eslint-disable-next-line @angular-eslint/directive-selector
                    selector: 'igx-column[email]',
                    providers: [{
                            provide: NG_VALIDATORS,
                            useExisting: IgxColumnEmailValidatorDirective,
                            multi: true
                        }],
                    standalone: true
                }]
        }] });
export class IgxColumnMinLengthValidatorDirective extends MinLengthValidator {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.2.4", ngImport: i0, type: IgxColumnMinLengthValidatorDirective, deps: null, target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "17.2.4", type: IgxColumnMinLengthValidatorDirective, isStandalone: true, selector: "igx-column[minlength]", providers: [{
                provide: NG_VALIDATORS,
                useExisting: IgxColumnMinLengthValidatorDirective,
                multi: true
            }], usesInheritance: true, ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.2.4", ngImport: i0, type: IgxColumnMinLengthValidatorDirective, decorators: [{
            type: Directive,
            args: [{
                    // eslint-disable-next-line @angular-eslint/directive-selector
                    selector: 'igx-column[minlength]',
                    providers: [{
                            provide: NG_VALIDATORS,
                            useExisting: IgxColumnMinLengthValidatorDirective,
                            multi: true
                        }],
                    standalone: true
                }]
        }] });
export class IgxColumnMaxLengthValidatorDirective extends MaxLengthValidator {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.2.4", ngImport: i0, type: IgxColumnMaxLengthValidatorDirective, deps: null, target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "17.2.4", type: IgxColumnMaxLengthValidatorDirective, isStandalone: true, selector: "igx-column[maxlength]", providers: [{
                provide: NG_VALIDATORS,
                useExisting: IgxColumnMaxLengthValidatorDirective,
                multi: true
            }], usesInheritance: true, ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.2.4", ngImport: i0, type: IgxColumnMaxLengthValidatorDirective, decorators: [{
            type: Directive,
            args: [{
                    // eslint-disable-next-line @angular-eslint/directive-selector
                    selector: 'igx-column[maxlength]',
                    providers: [{
                            provide: NG_VALIDATORS,
                            useExisting: IgxColumnMaxLengthValidatorDirective,
                            multi: true
                        }],
                    standalone: true
                }]
        }] });
export class IgxColumPatternValidatorDirective extends PatternValidator {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.2.4", ngImport: i0, type: IgxColumPatternValidatorDirective, deps: null, target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "17.2.4", type: IgxColumPatternValidatorDirective, isStandalone: true, selector: "igx-column[pattern]", providers: [{
                provide: NG_VALIDATORS,
                useExisting: IgxColumPatternValidatorDirective,
                multi: true
            }], usesInheritance: true, ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.2.4", ngImport: i0, type: IgxColumPatternValidatorDirective, decorators: [{
            type: Directive,
            args: [{
                    // eslint-disable-next-line @angular-eslint/directive-selector
                    selector: 'igx-column[pattern]',
                    providers: [{
                            provide: NG_VALIDATORS,
                            useExisting: IgxColumPatternValidatorDirective,
                            multi: true
                        }],
                    standalone: true
                }]
        }] });
