// igniteui-angular/esm2022/lib/tabs/bottom-nav/bottom-nav.directives.mjs
import { Directive } from '@angular/core';
import * as i0 from "@angular/core";
export class IgxBottomNavHeaderLabelDirective {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.2.4", ngImport: i0, type: IgxBottomNavHeaderLabelDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "17.2.4", type: IgxBottomNavHeaderLabelDirective, isStandalone: true, selector: "igx-bottom-nav-header-label,[igxBottomNavHeaderLabel]", ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.2.4", ngImport: i0, type: IgxBottomNavHeaderLabelDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: 'igx-bottom-nav-header-label,[igxBottomNavHeaderLabel]',
                    standalone: true
                }]
        }] });
export class IgxBottomNavHeaderIconDirective {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.2.4", ngImport: i0, type: IgxBottomNavHeaderIconDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "17.2.4", type: IgxBottomNavHeaderIconDirective, isStandalone: true, selector: "igx-bottom-nav-header-icon,[igxBottomNavHeaderIcon]", ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.2.4", ngImport: i0, type: IgxBottomNavHeaderIconDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: 'igx-bottom-nav-header-icon,[igxBottomNavHeaderIcon]',
                    standalone: true
                }]
        }] });
