// igniteui-angular/esm2022/lib/directives/button/button-base.mjs
import { Directive, EventEmitter, HostBinding, HostListener, Inject, Input, Optional, Output, booleanAttribute } from '@angular/core';
import { DisplayDensityBase, DisplayDensityToken } from '../../core/density';
import { mkenum } from '../../core/utils';
import * as i0 from "@angular/core";
export const IgxBaseButtonType = /*@__PURE__*/ mkenum({
    Flat: 'flat',
    Contained: 'contained',
    Outlined: 'outlined'
});
export class IgxButtonBaseDirective extends DisplayDensityBase {
    /**
     * @hidden
     * @internal
     */
    onClick(ev) {
        this.buttonClick.emit(ev);
    }
    /**
     * @hidden
     * @internal
     */
    get disabledAttribute() {
        return this.disabled || null;
    }
    constructor(element, _displayDensityOptions) {
        super(_displayDensityOptions, element);
        this.element = element;
        this._displayDensityOptions = _displayDensityOptions;
        /**
         * Emitted when the button is clicked.
         */
        this.buttonClick = new EventEmitter();
        /**
         * Sets/gets the `role` attribute.
         *
         * @example
         * ```typescript
         * this.button.role = 'navbutton';
         * let buttonRole = this.button.role;
         * ```
         */
        this.role = 'button';
        /**
          * Enables/disables the button.
          *
          * @example
          * ```html
          * <button igxButton="fab" disabled></button>
          * ```
          */
        this.disabled = false;
    }
    /**
     * Returns the underlying DOM element.
     */
    get nativeElement() {
        return this.element.nativeElement;
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.2.4", ngImport: i0, type: IgxButtonBaseDirective, deps: [{ token: i0.ElementRef }, { token: DisplayDensityToken, optional: true }], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "16.1.0", version: "17.2.4", type: IgxButtonBaseDirective, inputs: { disabled: ["disabled", "disabled", booleanAttribute] }, outputs: { buttonClick: "buttonClick" }, host: { listeners: { "click": "onClick($event)" }, properties: { "attr.role": "this.role", "class.igx-button--disabled": "this.disabled", "attr.disabled": "this.disabledAttribute" } }, usesInheritance: true, ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.2.4", ngImport: i0, type: IgxButtonBaseDirective, decorators: [{
            type: Directive
        }], ctorParameters: () => [{ type: i0.ElementRef }, { type: undefined, decorators: [{
                    type: Optional
                }, {
                    type: Inject,
                    args: [DisplayDensityToken]
                }] }], propDecorators: { buttonClick: [{
                type: Output
            }], role: [{
                type: HostBinding,
                args: ['attr.role']
            }], onClick: [{
                type: HostListener,
                args: ['click', ['$event']]
            }], disabled: [{
                type: Input,
                args: [{ transform: booleanAttribute }]
            }, {
                type: HostBinding,
                args: ['class.igx-button--disabled']
            }], disabledAttribute: [{
                type: HostBinding,
                args: ['attr.disabled']
            }] } });
